packages/@lwc/wire-service/src/wiring.ts

```diff
@@ -233,26 +235,28 @@ export class WireEventTarget {
                 this._cmp[this._wireTarget] = value;
             }
             return false; // canceling signal since we don't want this to propagate
-        } else if ((evt as ComposableEvent).type === 'WireContextEvent') {
-            // NOTE: kill this hack
-            // we should only allow ValueChangedEvent
-            // however, doing so would require adapter to implement machinery
-            // that fire the intended event as DOM event and wrap inside ValueChangedEvent
+        } else if (evt instanceof LinkContextEvent) {
+            const { uid, callback } = evt;
+            // this must dispatch an underlying dom event on the element to connect to
+            // the provider, which uses an internal unique ID defined per adapter, and
+            // provides a callback used as a side channel communication between provider
+            // and consumer adapter.
+            const internalDomEvent = new CustomEvent(uid, {
+                bubbles: true,
+                composed: true,
+                // avoid leaking the callback function directly to prevent a side channel
+                // during the linking phase to the context provider.
+                detail(...args) {
+                    callback(...args);
+                },
+            });
+            this._cmp.dispatchEvent(internalDomEvent);
+            return false; // canceling signal since we don't want this to propagate
+        } else if (evt.type === 'WireContextEvent') {
+            // TODO: issue #1357 - remove this branch
```

Comment:
@jbenallen asked me to keep this around, they are using it today, until they move to the new mechanism, we can keep this around, added the TODO, so we can track that.